xmlua/c14n.lua

```diff
@@ -0,0 +1,237 @@
+local C14n = {}
+
+local libxml2 = require("xmlua.libxml2")
+local ffi = require("ffi")
+
+local DEFAULT_MODE = "C14N_EXCLUSIVE_1_0"
+
+local Attribute
+local AttributeDeclaration
+local CDATASection
+local Comment
+local Document
+local DocumentFragment
+local DocumentType
+local Element
+local ElementDeclaration
+local EntityDeclaration
+local NamespaceDeclaration
+local Notation
+local ProcessingInstruction
+local Text
+
+function C14n.lazy_load()
+  Attribute = require("xmlua.attribute")
+  AttributeDeclaration = require("xmlua.attribute-declaration")
+  Text = require("xmlua.text")
+  CDATASection = require("xmlua.cdata-section")
+  Comment = require("xmlua.comment")
+  Document = require("xmlua.document")
+  DocumentFragment = require("xmlua.document-fragment")
+  DocumentType = require("xmlua.document-type")
+  Element = require("xmlua.element")
+  ElementDeclaration = require("xmlua.element-declaration")
+  EntityDeclaration = require("xmlua.entity-declaration")
+  NamespaceDeclaration = require("xmlua.namespace-declaration")
+  Notation = require("xmlua.notation")
+  ProcessingInstruction = require("xmlua.processing-instruction")
+end
+
+
+
+local C14N_MODES = {
+  C14N_1_0            = ffi.C.XML_C14N_1_0,           -- Original C14N 1.0 spec
+  C14N_EXCLUSIVE_1_0  = ffi.C.XML_C14N_EXCLUSIVE_1_0, -- Exclusive C14N 1.0 spec
+  C14N_1_1            = ffi.C.XML_C14N_1_1,           -- C14N 1.1 spec
+}
+
+
+
+local C14N_MODES_LOOKUP = {} -- lookup by name or number, returns the number
+for name, number in pairs(C14N_MODES) do
+  C14N_MODES_LOOKUP[name] = number
+  C14N_MODES_LOOKUP[number] = number
+end
+
+
+
+-- list can be a string (space separated), an array of strings, or nil
+local function getNamespacePrefixArray(list)
+  local list = list or {}
+
+  if type(list) == "string" then
+    -- list is a string, assume it is the space separated PrefixList attribute, split it
+    local list_str = list
+    list = {}
+    list_str:gsub("([^%s]+)", function(cap) list[#list+1] = cap end)
+  end
+
+  if #list == 0 then
+    return nil
+  end
+
+  local result = ffi.new('xmlChar*[?]', #list+1)
+  local refs = {}
+  for i, prefix in ipairs(list) do
+    local c_ns = ffi.new("unsigned char[?]", #prefix+1, prefix)
+    ffi.copy(c_ns, prefix)
+    result[i-1] = c_ns
+    refs[i] = c_ns -- hold on to refs to prevent GC while in use
+  end
+  result[#list] = nil
+
+  return ffi.gc(result, function(ptr)
+    refs = nil -- release references, so they can be GC'ed
+  end)
+end
+
+
+
+local function getNodesList(nodes)
+  if (not nodes) or #nodes == 0 then
+    return nil
+  end
+
+  local nodeTab = ffi.new("xmlNodePtr[?]", #nodes)
+  for i = 1, #nodes do
+      nodeTab[i - 1] = nodes[i] -- FFI side is 0 indexed
+  end
+
+  local set = ffi.new("xmlNodeSet")
+  set.nodeNr = #nodes
+  set.nodeMax = #nodes
+  set.nodeTab = nodeTab
```

Comment:
Can this prevent GC `nodeTab` while `set` is alive?

Reply:
you mean that if the function exits, `nodeTab` goes out of scope, and is GC'ed. Which then leads to `set.nodeTab` to become invalid (pointing to already released memory).

Not sure about this, not an FFI expert...

Reply:
I tried the following:

```lua
local xmlua = require("xmlua")
local ffi = require("ffi")
local nodeTab = ffi.gc(ffi.new("xmlNodePtr[?]", 3), function(node) print("freed")end)
local set = ffi.new("xmlNodeSet")
set.nodeTab = nodeTab
nodeTab = nil
print("before GC")
collectgarbage()
print("after GC")
```

Result:

```text
before GC
freed
after GC
```

It shows that this code isn't safe. Could you return not only `set` but also `nodeTab` and keep referring them in caller?

Reply:
this works without having to keep track of the `nodeTab` manually:
```lua
local xmlua = require("xmlua")
local ffi = require("ffi")

local function create_nodeset()
  local nodeTab = ffi.gc(ffi.new("xmlNodePtr[?]", 3), function(node) print("freed")end)
  local set = ffi.new("xmlNodeSet")
  set.nodeTab = nodeTab
  return ffi.gc(set, function(ptr)
    nodeTab = nil -- release references, so they can be GC'ed
  end)
end

nodeset = create_nodeset()
print("before GC")
collectgarbage()
collectgarbage()
print("after GC 1")
nodeset = nil
collectgarbage()
collectgarbage()
print("after GC 2")
```

returns
```
before GC
after GC 1
freed
after GC 2
```

Reply:
fixed it.

Reply:
Ah, we have already used this approach in other places.
Great!